Add State overrides to eth_call

maxencerb opened the pull request:
## What

The goal of this PR is to add state overrides to call and `eth_call` EIP1193 types. These overrides are helpful in many cases. It modifies the state of selected accounts before the transaction. It is a default feature in geth. It allows to modify:

- The balance of the account
- The code on a given account
- The state (key-value mapping representation of the storage) either on top of the account storage or completely override it.

## Why

I only added the rpc schema and changed the "low-level" call function. But the first case scenario that comes in mind is the approve and swap transaction simulation.

If you know the mapping key for the allowances, you could change the allowance to the max allowance before a `simulateContract` call. This would allow a cheap onchain simulation of the rates you could get.

```ts
const { result } = await publicClient.simulateContract({
  address: '0xFBA3912Ca04dd458c843e2EE08967fC04f3579c2',
  abi: mangroveABI,
  functionName: 'marketOrderByTick',
  account,
  stateOverride: {
    // modify the state of the USDT contract
    [USDT_ADDRESS]: {
      // Allows to modify storage while keeping other storage intact
      stateDiff: {
        // modify the storage slot corresponding to allowance
        [userAllowanceSlot]: "0xffffffffff...",
      }
    }
  }
})
```

<!-- start pr-codex -->

---

## PR-Codex overview
### Focus of this PR:
This PR focuses on adding the `stateOverride` property to various functions (`call`, `simulateContract`, `readContract`, and `multicall`) and introducing related types and error classes.

### Detailed summary:
- Added `stateOverride` property to `call`, `simulateContract`, `readContract`, and `multicall` functions.
- Added `StateOverride` type defining state overrides for `eth_call` method.
- Added `InvalidBytesLengthErrorType` and `InvalidBytesLengthError` classes.
- Added `RpcStateOverride` and `RpcStateMapping` types.
- Added `StateMapping` and `StateOverride` types.
- Added `AccountStateConflictErrorType`, `AccountStateConflictError`, `StateAssignmentConflictErrorType`, and `StateAssignmentConflictError` classes.
- Updated various files with the new changes.

> The following files were skipped due to too many changes: `src/errors/stateOverride.ts`, `src/actions/public/multicall.test.ts`, `site/pages/docs/actions/public/call.md`, `src/actions/public/call.ts`, `site/pages/docs/contract/simulateContract.md`, `src/actions/public/call.test.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

<!-- end pr-codex -->

jxom commented on src/actions/public/call.ts:
```diff
@@ -66,6 +71,27 @@ export type FormattedCall<
+  nonce?: number
+  code?: Hex
+} & (
+  | {
+      /** Fake key-value mapping to override all slots in the account storage before executing the call. */
+      state?: StateMapping
+      stateDiff?: never
+    }
+  | {
+      /** Fake key-value mapping to override individual slots in the account storage before executing the call. */
+      stateDiff?: StateMapping
+      state?: never
+    }
```
We can use our `OneOf` type util here.

```suggestion
  OneOf<
    | {
      /** Fake key-value mapping to override all slots in the account storage before executing the call. */
      state?: StateMapping
    }
    | {
      /** Fake key-value mapping to override individual slots in the account storage before executing the call. */
      stateDiff?: StateMapping
    }
  }>
```

jxom commented on src/actions/public/call.ts:
```diff
@@ -169,7 +200,7 @@ export async function call<TChain extends Chain | undefined>(
     } as TransactionRequest) as TransactionRequest
 
-    if (batch && shouldPerformMulticall({ request })) {
+    if (!rawStateOverride && batch && shouldPerformMulticall({ request })) {
```
Might read better after the `batch` condition. 

"If we have **batching enabled**, and blah blah blah"

```suggestion
    if (batch && shouldPerformMulticall({ request }) && !rawStateOverride) {
```

jxom reviewed (commented):
Seems chill. Just a couple of comments!

tmm commented on src/types/misc.ts:
```diff
@@ -17,3 +18,25 @@ export type CompactSignature = {
+  [slots: Hex]: Hex
+}
+
+export type RawAccountStateOverride = Partial<{
+  /** Fake balance to set for the account before executing the call. <32 bytes */
+  balance: Hex
+  /** Fake nonce to set for the account before executing the call. <8 bytes */
+  nonce: Hex
+  /** Fake EVM bytecode to inject into the account before executing the call. */
+  code: Hex
+  /** Fake key-value mapping to override all slots in the account storage before executing the call. */
+  state: StateMapping
+  /** Fake key-value mapping to override individual slots in the account storage before executing the call. */
+  stateDiff: StateMapping
+}>
```
Better to be explicit about it.

```suggestion
export type RawAccountStateOverride = {
  /** Fake balance to set for the account before executing the call. <32 bytes */
  balance?: Hex
  /** Fake nonce to set for the account before executing the call. <8 bytes */
  nonce?: Hex
  /** Fake EVM bytecode to inject into the account before executing the call. */
  code?: Hex
  /** Fake key-value mapping to override all slots in the account storage before executing the call. */
  state?: StateMapping
  /** Fake key-value mapping to override individual slots in the account storage before executing the call. */
  stateDiff?: StateMapping
}
```

jxom commented on src/types/misc.ts:
Maybe these changes should belong in `types/rpc.ts`, and `RpcAccountStateOverride` instead of `RawAccountStateOverride`?

jxom commented on src/actions/public/call.ts:
```diff
@@ -66,6 +71,27 @@ export type FormattedCall<
   TChain extends Chain | undefined = Chain | undefined,
 > = FormattedTransactionRequest<TChain>
 
+export type AccountStateOverride = {
+  balance?: bigint
+  nonce?: number
+  code?: Hex
+} & (
+  | {
+      /** Fake key-value mapping to override all slots in the account storage before executing the call. */
+      state?: StateMapping
+      stateDiff?: never
+    }
+  | {
+      /** Fake key-value mapping to override individual slots in the account storage before executing the call. */
+      stateDiff?: StateMapping
+      state?: never
+    }
+)
+
+export type StateOverride = {
+  [account: Address]: AccountStateOverride
+}
```
Maybe we put these into `types/stateOverride.ts`?

maxencerb commented:
I made some changes, added tests for `readContract`, `multicall`, and `call`, added some error handling. 

Also added API docs on `simulateContract`, `readContract`, `multicall`, and `call` and added a usage example on `simulateContract`.

jxom reviewed (commented):
It's looking good! If we can get coverage on these lines, then I'm happy to merge!

`bun run test:cov public/call.test`

<img width="446" alt="Screenshot 2024-02-02 at 1 02 58 pm" src="https://github.com/wevm/viem/assets/7336481/254a7c35-9c07-4dcb-a9e4-b4764551ce3a">

maxencerb commented:
I added coverage to these lines on `call.test.ts`.
